binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/InteractiveQueryService.java

```diff
@@ -104,28 +114,51 @@ public <T> T getQueryableStore(String storeName, QueryableStoreType<T> storeType
 				LOG.warn("Store (" + storeName + ") could not be found in Streams context, falling back to all known Streams instances");
 			}
 
+			// Find all apps that know about the store
+			Map<KafkaStreams, T> candidateStores = new HashMap<>();
 			for (KafkaStreams kafkaStreamApp : kafkaStreamsRegistry.getKafkaStreams()) {
 				try {
-					return getStateStoreFromKafkaStreams(kafkaStreamApp, storeName, storeType);
+					candidateStores.put(kafkaStreamApp, kafkaStreamApp.store(storeQueryParams));
 				}
 				catch (Exception ex) {
 					throwable = ex;
 				}
 			}
+
+			// Store exists in a single app - no further resolution required
+			if (candidateStores.size() == 1) {
+				return candidateStores.values().stream().findFirst().get();
+			}
+
+			// If the store is in multiple streams apps - discard any apps that do not actually have the store
+			if (candidateStores.size() > 1) {
+
+				candidateStores = candidateStores.entrySet().stream()
+						.filter((e) -> storeActuallyAvailable(e.getKey(), storeName))
+						.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+				if (candidateStores.size() == 1) {
+					return candidateStores.values().stream().findFirst().get();
+				}
+
+				throwable = (candidateStores.size() == 0) ?
+						new UnknownStateStoreException("Store (" + storeName + ") not available to Streams instance") :
+						new InvalidStateStoreException("Store (" + storeName + ") available to more than one Streams instance");
+
+			}
 			throw new IllegalStateException("Error retrieving state store: " + storeName, throwable);
 		});
 	}
 
-	private <T> T getStateStoreFromKafkaStreams(KafkaStreams kafkaStreams, String storeName, QueryableStoreType<T> storeType) {
-		// Check KafkaStreams app knows about the state store
-		T store = kafkaStreams.store(StoreQueryParameters.fromNameAndType(storeName, storeType));
-
-		// Check KafkaStreams app actually has the state store
-		if (kafkaStreams.streamsMetadataForStore(storeName).stream()
-				.noneMatch((sm) -> sm.stateStoreNames().contains(storeName))) {
-			throw new UnknownStateStoreException("Store (" + storeName + ") not available to Streams instance");
+	private boolean storeActuallyAvailable(KafkaStreams kafkaStreams, String storeName) {
+		try {
+			TopologyMetadata topologyMetadata = (TopologyMetadata) ReflectionUtils.getField(topologyMetadataField, kafkaStreams);
+			return !topologyMetadata.sourceTopicsForStore(storeName, null).isEmpty();
+		}
+		catch (Exception ex) {
+			LOG.error("Unable to determine if store (" + storeName + ") is available to the app due to: " + ex.getMessage(), ex);
```

Comment:
When/if this fails, we will not surface exception but rather log and let resolve process continue.